### DISCUSSION TITLE
How do I assert the right type for a populated entity?

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [mikro-orm/mikro-orm#2940] [Q&A] [answered]
The code below shows a type of `{ id: string; } & Reference<User>` when it should be of type `User` since I populated it... any ideas?

``` typescript
const entities = await em
  .createQueryBuilder(ParentEntity)
  .select('*')
  .where(whereOptions);

await req.entityManager.populate(entities, ['user']);

for (let i = 0; i < entities.length; i += 1) {
  const entity = entities[i];
  const { someProperty } = entity.user; // Property 'someProperty' does not exist on type '{ id: string; } & Reference<User>'
}
```

I initially tried `unwrap` but it doesn't exist on the user... if I assert it to `any` all the properties I need _**are**_ populated, so it's just a type issue.

I'm doing exactly [what the docs suggest](https://mikro-orm.io/docs/nested-populate#populating-already-loaded-entities), so I'm unsure of what I'm doing wrong 🤔

**Comment by johnkahn** [comment]
Quick solution: call `.unwrap()` on the nested entity you populated. But this would introduce a problem where you could remove the `populate()` call and Typescript wouldn't throw an error because it doesn't relate the two calls.

A better one IMO: call `.load()` on it and await it in the loop instead of calling `populate`. That's basically what populate is doing AFAIK anyway, but would prevent the above problem from happening.

**Reply by SpencerKaiser** [reply]
`unwrap` is `undefined`... that was my first approach and then I got runtime errors because none of the helper methods are present.

And I could use load but I assumed the `populate` call was loading all of them instead of making individual calls.

**Reply by johnkahn** [reply]
Hm, yeah I think you're right, it does batch them

**Reply by johnkahn** [reply]
Does populate return the entities with loaded fields?

**Reply by SpencerKaiser** [reply]
Literally everything is there, it's just the type that goofed up 😞 trying to assert it as `User` throws a type warning but if I assert it as an `any` everything works perfectly

**Reply by johnkahn** [reply]
Yeah, Typescript won't modify types like that, so it has to be put into a new variable with the correct type. So if `populate()` returns the loaded entities I would assume that it would be the correct type. I'll try to reproduce this in another project so I can play around with it and see what I can come up with for a better pattern.

**Reply by SpencerKaiser** [reply]
@B4nan any ideas??

**Comment by B4nan** [comment, marked as answer]
>  when it should be of type User since I populated it... any ideas?

It should _not_ work like that, not sure why you think so. You define the relation as wrapped reference, and it will be _always_ wrapped reference. 

> I initially tried unwrap but it doesn't exist on the user..

If you have user instance there, you have wrong entity definition, you are missing `wrappedReference: true` in the property options.

> I'm doing exactly [what the docs suggest](https://mikro-orm.io/docs/nested-populate#populating-already-loaded-entities), so I'm unsure of what I'm doing wrong 🤔

Not really, you are talking about wrapped references, this section has nothing to do with them. 

https://mikro-orm.io/docs/entity-references#better-type-safety-with-referencet-wrapper

**Reply by SpencerKaiser** [reply]
> It should not work like that, not sure why you think so. You define the relation as wrapped reference, and it will be always wrapped reference.

I think the issue on my end is a misunderstanding of what `wrappedReference` really means. I really struggled while reading through the docs because I couldn't find the high-level description that communicates exactly what it means to have something be wrapped/unwrapped. The combo of your context here and this statement finally made it click:
> You can overcome this issue by using the Reference<T> wrapper. It simply wraps the entity, defining load(): Promise<T> method that will first lazy load the association if not already available. 

I would love to see that chunk get more emphasis in the docs because I think it's tremendously important and right now it's just a little bit of text between sections 😉

> Not really, you are talking about wrapped references, this section has nothing to do with them.

Yea this was because of my misunderstanding with how the entity definition had major implications on the types of things down the line (which makes perfect sense, but it was difficult to grasp) 😅 

📣 For anyone else finding this, @B4nan's suggestion about `wrappedReference: true` is what fixed the issue for me. I made this change to the `ManyToOne` relationship in my entity declaration for `ParentEntity`:
``` typescript
@ManyToOne(() => User, { wrappedReference: true })
user: IdentifiedReference<User>;
```

Thanks @B4nan, I really appreciate your quick response as always!!! 🚀

**Reply by B4nan** [reply]
docs could be indeed more verbose, PR welcome :] its always hard to find time for that...
